### PR TITLE
Replace @completionHandlerAsync with @available(*, renamed:)

### PR DESCRIPTION
**Pull request by bnbarham**
Instead of a new attribute `@completionHandlerAsync`, allow the use of
the existing `renamed` parameter of `@available` to specify the
asynchronous alternative of a synchronous function.

No errors will be output from invalid names as `@completionHandlerAsync`
had, but if a function is correctly matched then it will be used to
output warnings when using the synchronous function in an asynchronous
context (as before).

Resolves rdar://80612731

### CODE REVIEW

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
Few main changes from the commit in the other PR:
  - No `Decl::getRenamedDecl`, it only makes sense in the context of a particular attribute
  - `getAsyncAlternative` uses the first-in-source `@available` that has no platform and a `renamed`, or the first with a `renamed` otherwise (regardless of platform)
  - `findPotentialCompletionHandlerParam` ignores default parameters in the async function if the argument labels don't match. Added a bunch of tests for that case

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test Windows platform

**Review by hamishknight — APPROVED** [review]
Looks good to me!

File: lib/AST/Attr.cpp
Comment (bnbarham) [reply lib/AST/Attr.cpp:1466]: Yeah that's probably fair. It only has the one callsite from the clang importer at the moment anyway, which should be implicit.

File: lib/AST/Decl.cpp
Comment (bnbarham) [reply lib/AST/Decl.cpp:7140]: Ah thanks, really nice catch. Not sure how I missed that comment, or `isDefaultArgument` for that matter. I think I was looking for "has...".

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test

File: lib/AST/Decl.cpp
Comment (hamishknight) [reply lib/AST/Decl.cpp]: Yeah we could definitely do better on the naming of those, `isDefaultArgument` should definitely be `hasDefaultArgument`, and `hasDefaultExpr` should maybe be something like `hasParsedDefaultExpr`

**Comment by bnbarham** [comment]
Ah, someone else got module format version number change in before I could merge. Ah well. Last tests passed, will just do a smoke test so I can merge.

@swift-ci smoke test

**Comment by bnbarham** [comment]
@etcwilde I'm going to merge this, but let me know if you think it needs any changes.

**Pull request merged**
